nntrainer/tensor/cpu_backend/x86/avx2_impl.cpp

```diff
@@ -12,12 +12,298 @@
+                                 0x3f75fed7U, 0x3f7504f3U, 0x3f75672aU,
+                                 0x3f7744fdU, 0x3f7ac0c7U};
+    return t;
+  }
+};
+
+#endif
+
+template <unsigned N_> // requires PowerOfTwo<N_>
+inline constexpr auto exp2_table_v = Exp2Table<N_>::calculate();
+
+template <unsigned N_, typename Ty_> // requires PowerOfTwo<N_>
+inline constexpr auto exp2_table_type_v = Exp2Table<N_, Ty_>::calculate();
+
+// Scalar version of expf() approximation with 3-th deg polynominal of
+// fractional part
+//
+// The error with regards to std::expf less than 5e-6
+// It is valid in range [-87, +88.37) - not handling +INF, NaN etc.
+// The function domain is clamped to valid function range.
+//
+// The strategy picked is to approximate exp as 2^K*2*F
+template <unsigned N_, typename = std::enable_if_t<concept17_PowerOfTwo<N_>>>
+[[nodiscard]] _nnt_ATTR_ALWAYS_INLINE _nnt_ATTR_FLATTEN inline float
+approx_exp_exp2_lookup(float x) noexcept {
+  constexpr static unsigned N_MASK = uint32_t(N_ - 1U);
+  constexpr static unsigned FLT_MANTISSA_BITS =
+    std::numeric_limits<float>::digits - 1U;
+
+  x = std::max(x, EXP_ARG_MIN);
+  x *= float(0x1.0p1 / numbers::ln2_v<double> * N_);
+  x = std::min(x, float(EXP_ARG_MAX / numbers::ln2_v<double> * N_));
+
+  // Round nearest and convert integer part to an int (std::modf)
+  // NB: This way doesn't handle ties even.
+  auto x_int = x + 0x1.8p23f;
+  auto x_uint = compat_bit_cast<uint32_t>(x_int);
+  x_int -= 0x1.8p23f;
+  auto x_frac = x - x_int;
+
+  auto s_int = exp2_table_v<N_>[x_uint & N_MASK];
+  auto x_uint_shifted = x_uint
+                        << (FLT_MANTISSA_BITS - constexpr_popcount(N_MASK));
+  auto s_int_2 = s_int + x_uint_shifted;
+  auto s = compat_bit_cast<float>(s_int_2);
+
+  // Polynominal of form C0*x^3 + C1*x^2 + C2*x^1 + 1.0
+  static constexpr float poly_4[] = {0x1.c6af84b912394p-5 / N_ / N_ / N_,
+                                     0x1.ebfce50fac4f3p-3 / N_ / N_,
+                                     0x1.62e42ff0c52d6p-1 / N_};
+
+  auto q0 = std::fma(x_frac, poly_4[0], poly_4[1]);
+  auto x_frac_pow2 = x_frac * x_frac;
+  auto q2 = x_frac * poly_4[2]; // not adding +1.0
+
+  x = std::fma(q0, x_frac_pow2, q2);
+
+  return std::fma(x, s, s); // NB: (handles (x+1) by addition of s)
+}
+
+// Vectorized version of above
+template <unsigned N_, typename = std::enable_if_t<concept17_PowerOfTwo<N_>>>
+_nnt_ATTR_ALWAYS_INLINE _nnt_ATTR_FLATTEN inline auto _nnt_CC_VECTORCALL
+avx2_approx_exp_e2lookup(__m256 xs) noexcept {
+
+  constexpr static uint32_t N_MASK = uint32_t(N_ - 1U);
+  constexpr static auto EXP2_TBL = exp2_table_v<N_>;
+  constexpr static unsigned MANTISSA_BITS =
+    std::numeric_limits<float>::digits - 1;
+
+  // Ensure arg in range [exp_arg_min, exp_arg_max]
+  xs = _mm256_max_ps(xs, _mm256_set1_ps(EXP_ARG_MIN));
+  // Would clamp to EXP_ARG_MAX but we move it after multiplication for IPC
+  xs = _mm256_min_ps(xs, _mm256_set1_ps(EXP_ARG_MAX));
```

Comment:
Should be commented out, clamp to upper bound is handled below after multiplication.